pkg/apis/migration/v1alpha1/condition.go

```diff
@@ -40,110 +51,193 @@ func (r *Condition) Update(other Condition) {
+	if r.List == nil {
+		r.List = []Condition{}
 	}
 	condition.staged = true
-	_, found := r.FindCondition(condition.Type)
+	found := r.FindCondition(condition.Type)
 	if found == nil {
 		condition.LastTransitionTime = metav1.NewTime(time.Now())
-		r.Conditions = append(r.Conditions, condition)
+		r.List = append(r.List, condition)
 	} else {
 		found.Update(condition)
 	}
 }
 
 // Delete conditions by type.
 func (r *Conditions) DeleteCondition(cndTypes ...string) {
-	if r.Conditions == nil {
+	if r.List == nil {
 		return
 	}
 	kept := []Condition{}
 	for _, name := range cndTypes {
-		for index := range r.Conditions {
-			condition := r.Conditions[index]
+		for i := range r.List {
+			condition := r.List[i]
 			if condition.Type != name {
 				kept = append(kept, condition)
 			}
 		}
 	}
-	r.Conditions = kept
+	r.List = kept
 }
 
-// Mark all conditions as not staged.
-func (r *Conditions) UnstageConditions() {
-	if r.Conditions == nil {
-		return
+// The collection has ALL of the specified conditions.
+func (r *Conditions) HasCondition(types ...string) bool {
+	if r.List == nil {
+		return false
 	}
-	for index := range r.Conditions {
-		condition := &r.Conditions[index]
-		condition.staged = false
+	for _, cndType := range types {
+		condition := r.FindCondition(cndType)
+		if condition == nil || condition.Status != True {
+			return false
+		}
+		if r.staged && !condition.staged {
+			return false
+		}
 	}
+
+	return len(types) > 0
 }
 
-// Delete un-staged conditions.
-func (r *Conditions) DeleteUnstagedConditions() {
-	if r.Conditions == nil {
-		return
+// The collection contains any conditions with category.
+func (r *Conditions) HasConditionCategory(names ...string) bool {
+	if r.List == nil {
+		return false
 	}
-	kept := []Condition{}
-	for index := range r.Conditions {
-		condition := r.Conditions[index]
-		if condition.staged {
-			kept = append(kept, condition)
+	catSet := map[string]bool{}
+	for _, name := range names {
+		catSet[name] = true
+	}
+	for _, condition := range r.List {
+		_, found := catSet[condition.Category]
+		if !found || condition.Status != True {
+			continue
+		}
+		if r.staged && !condition.staged {
+			continue
 		}
+		return true
 	}
-	r.Conditions = kept
+
+	return false
+}
+
+// The collection contains an `Error` condition.
+func (r *Conditions) HasErrorCondition(category ...string) bool {
+	return r.HasConditionCategory(Error)
+}
+
+// The collection contains a `Warn` condition.
+func (r *Conditions) HasWarnCondition(category ...string) bool {
+	return r.HasConditionCategory(Error)
+}
+
+// The collection contains a `Ready` blocker condition.
+func (r *Conditions) HasBlockerCondition() bool {
+	return r.HasConditionCategory(Error, Warn)
 }
 
 // Set `Ready` condition.
 func (r *Conditions) SetReady(ready bool, message string) {
 	if ready {
 		r.SetCondition(Condition{
-			Type:    Ready,
-			Status:  True,
-			Message: message,
+			Type:     Ready,
+			Status:   True,
+			Category: Required,
```

Comment:
So do we check that at least one condition with `Category: Required` is set on a resource before proceeding with Migration? I haven't seen this in your PR but will look further.

Reply:
I'm conflicted on our "good" category being called `Required`. Would we ever want to show a "good" condition that isn't "required"?